src/action.cpp

```diff
@@ -2323,14 +2330,16 @@ static void actionDroidBase(DROID *psDroid, DROID_ACTION_DATA *psAction)
 				turnOffMultiMsg(false);
 			}
 		}
-		else if (order->type != DORDER_HOLD && secondaryGetState(psDroid, DSO_HALTTYPE) != DSS_HALT_HOLD) // approach closer?
+		else if (order->type != DORDER_HOLD 
+				&& secondaryGetState(psDroid, DSO_HALTTYPE) != DSS_HALT_HOLD
+				&& !actionIsInRange) // approach closer?
 		{
 			psDroid->action = DACTION_MOVETOATTACK;
 			turnOffMultiMsg(true);
 			moveDroidTo(psDroid, psAction->psObj->pos.x, psAction->psObj->pos.y);
 			turnOffMultiMsg(false);
 		}
-		else if (order->type != DORDER_HOLD && secondaryGetState(psDroid, DSO_HALTTYPE) == DSS_HALT_HOLD)
+		else if (actionIsInRange ||(order->type != DORDER_HOLD && secondaryGetState(psDroid, DSO_HALTTYPE) == DSS_HALT_HOLD))
```

Comment:
&&?

Reply:
`||` is correct
If action is in range, whatever is `DORDER_*`/ `DSO_*`, we can just start shooting right now. 

Note how the condition above (`// approach closer?`) didn't actually check if target is `out-of-range` before giving `DACTION_MOVETOATTACK`; but now it does.